docs/async-await-proposal.md

```diff
@@ -0,0 +1,442 @@
+
+  func expand(
+    _ request: Echo_EchoRequest,
+    callOptions: CallOptions?,
+    handler: @escaping (Echo_EchoResponse) -> Void
+  ) -> ServerStreamingCall<Echo_EchoRequest, Echo_EchoResponse>
+
+  func collect(
+    callOptions: CallOptions?
+  ) -> ClientStreamingCall<Echo_EchoRequest, Echo_EchoResponse>
+
+  func update(
+    callOptions: CallOptions?,
+    handler: @escaping (Echo_EchoResponse) -> Void
+  ) -> BidirectionalStreamingCall<Echo_EchoRequest, Echo_EchoResponse>
+}
+```
+
+### Existing example client usage
+
+Here is an example use of the client, making a bidirectional streaming call to
+`update`:
+
+```swift
+// Update is a bidirectional streaming call; provide a response handler.
+let update = client.update { response in
+  print("update received: \(response.text)")
+}
+
+// Send a bunch of messages to the service.
+for word in ["boyle", "jeffers", "holt"] {
+  let request = Echo_EchoRequest.with { $0.text = word }
+  update.sendMessage(request, promise: nil)
+}
+
+// Close the request stream.
+update.sendEnd(promise: nil)
+
+// wait() for the call to terminate
+let status = try update.status.wait()
+print("update completed with status: \(status.code)")
+```
+
+This API also exposes a number incidental types and patterns that the user need
+concern themselves with that are not specific to their application:
+
+1. It exposes the NIO types to the user, allowing the provision of an
+   `EventLoopPromise` when sending messages and requiring the use of
+   `EventLoopFuture` to obtain the `status` of the call.
+2. Code does not read in a straight line due to the need to provide a completion
+   handler when making the call.
+
+## Proposed solution
+
+### Proposed server API
+
+We propose generating the following new protocol which the user must conform to in
+order to implement the server:
+
+```swift
+/// To build a server, implement a class that conforms to this protocol.
+public protocol Echo_AsyncEchoProvider: CallHandlerProvider {
+  var interceptors: Echo_EchoServerInterceptorFactoryProtocol? { get }
+
+  /// Immediately returns an echo of a request.
+  func get(
+    request: Echo_EchoRequest,
+    context: AsyncServerCallContext
+  ) async throws -> Echo_EchoResponse
+
+  /// Splits a request into words and returns each word in a stream of messages.
+  func expand(
+    request: Echo_EchoRequest,
+    responseStreamWriter: AsyncResponseStreamWriter<Echo_EchoResponse>,
+    context: AsyncServerCallContext
+  ) async throws
+
+  /// Collects a stream of messages and returns them concatenated when the caller closes.
+  func collect(
+    requests: GRPCAsyncStream<Echo_EchoRequest>,
+    context: AsyncServerCallContext
+  ) async throws -> Echo_EchoResponse
+
+  /// Streams back messages as they are received in an input stream.
+  func update(
+    requests: GRPCAsyncStream<Echo_EchoRequest>,
+    responseStreamWriter: AsyncResponseStreamWriter<Echo_EchoResponse>,
+    context: AsyncServerCallContext
+  ) async throws
```

Comment:
This boils down to opinions™ but worth considering that this still allows the error of forgetting to complete the response stream etc.

We went through the same design dance way back when with akka-grpc and went with accepting and emitting a stream (`Source` in Akka's wording); https://doc.akka.io/docs/akka-grpc/current/server/walkthrough.html#implementing-the-service

It may be that Swift is still too rigid to allow that... we may first need the ability to say `some AsyncSequence where Element == Echo_EchoResponse` to be honest... 🤔 I believe this feature is coming "soon". 

But just throwing it out there for consideration.

Reply:
(I amend this opinion a bit below)

Reply:
I agree that we would like `some AsyncSequence where Element == ...`. George and I were mulling over the same as this would allow us to use opaque return type without having to provide the `GRPCAsyncStream` wrapper in this proposal.

Reply:
I added a note about the desire for an opaque `AsyncSequence` return type in https://github.com/grpc/grpc-swift/pull/1231/commits/da20f7d14a74af1ca521a75ffab171600596944b.